pyogrio/_io.pyx

```diff
@@ -1390,19 +1423,28 @@ def ogr_open_arrow(
         # make sure layer is read from beginning
         OGR_L_ResetReading(ogr_layer)
 
-        if not OGR_L_GetArrowStream(ogr_layer, &stream, options):
-            raise RuntimeError("Failed to open ArrowArrayStream from Layer")
+        # allocate the stream struct and wrap in capsule to ensure clean-up on error
+        if not return_pyarrow:
+            capsule = alloc_c_stream(&stream)
+        else:
+            stream = <ArrowArrayStream*> malloc(sizeof(ArrowArrayStream))
 
-        stream_ptr = <uintptr_t> &stream
+        if not OGR_L_GetArrowStream(ogr_layer, stream, options):
+            if return_pyarrow:
+                free(stream)
+            raise RuntimeError("Failed to open ArrowArrayStream from Layer")
 
         if skip_features:
             # only supported for GDAL >= 3.8.0; have to do this after getting
             # the Arrow stream
             OGR_L_SetNextByIndex(ogr_layer, skip_features)
 
-        # stream has to be consumed before the Dataset is closed
-        import pyarrow as pa
-        reader = pa.RecordBatchStreamReader._import_from_c(stream_ptr)
+        if return_pyarrow:
+            import pyarrow as pa
```

Comment:
This may raise an exception, in which case you want to `free(ptr)`. You also want to free the pointer after `_import_from_c` anyway (the structure's contents will have been moved, but the structure itself must still be deallocated anyway).

I think it's easier and safer to always create the capsule, since that will ensure correct deallocation.

Reply:
Ah, yes, good catch.

I wasn't yet creating the capsule for pyarrow, because `_import_from_c_capsule` only exists in recent pyarrow. But, I assume I can still create the capsule and just pass the pointer to `pa.RecordBatchReader._import_from_c`. Once that is called, pyarrow will ensure to move the struct, such that the capsule destructor (even the though the capsule itself wasn't passed to pyarrow) will not call the release callback. 

That indeed also allows to clean up the explicit `free(stream)` call in the `finally` clause, as at the point the capsule will go out of scope and free it instead.